### ISSUE TITLE
Pass.FullScreenQuad causes memory leak

### ISSUE THREAD

**Issue by yushijinhun** [mrdoob/three.js#20346]
Each time `Pass.FullScreenQuad` is used in a new WebGLRenderer, a `dispose` listener is attached to `geometry`. But when WebGLRenderer is disposed, `geometry` is not disposed, therefore the reference to the `dispose` listener is retained.

https://github.com/mrdoob/three.js/blob/b0a1e8ca7863aa399c31692fa5b76c40feca2485/examples/jsm/postprocessing/Pass.js#L37-L40

The listener holds a reference to `info` and `bindingStates`, which further refers to many other objects.

https://github.com/mrdoob/three.js/blob/4944e56c06ed284abe11b7654348fb1ea3ac04d4/src/renderers/webgl/WebGLGeometries.js#L5

![image](https://user-images.githubusercontent.com/8373027/93076018-4548c100-f6b9-11ea-8e76-8c61a386075e.png)

![image](https://user-images.githubusercontent.com/8373027/93077190-0e73aa80-f6bb-11ea-861e-8924cc88000f.png)

**Comment by Mugen87** [comment]
`FullScreenQuad` provides a `dispose()` method. And all FX passes should have a `fsQuad` property. Have you considered to call the method over this property?

**Comment by yushijinhun** [comment]
`fsQuad.dispose()` solves the memory leak problem. But I think there is something wrong with the method.

Because `FullScreenQuad._mesh.geometry` is a shared object, calling `fsQuad.dispose()` will also affect other `FullScreenQuad` instances. I don't know whether this behavior will cause problems, but it seems quite strange.

**Comment by Mugen87** [comment]
It will not cause a problem on app level. The call of `dispose()` only means the respective `WebGLBuffer` has to be recreated. Since the related buffer is so small, it don't think this will cause much of an overhead.

/cc @Oletus

**Comment by gkjohnson** [comment]
@Mugen87 I think this is an issue with all geometry that might get shared between multiple renderers.

In `WebGLGeometries` a [dispose callback](https://github.com/mrdoob/three.js/blob/4944e56c06ed284abe11b7654348fb1ea3ac04d4/src/renderers/webgl/WebGLGeometries.js#L60) is added to the geometry for that renderer which is only removed [when dispose is fired from the geometry](https://github.com/mrdoob/three.js/blob/4944e56c06ed284abe11b7654348fb1ea3ac04d4/src/renderers/webgl/WebGLGeometries.js#L27). Those callbacks for that renderer specifically are never removed otherwise even when [WebGLRenderer.dispose](https://github.com/mrdoob/three.js/blob/2299282085b48ffa9174cce5542c5041bd0ed7c0/src/renderers/WebGLRenderer.js#L583-L599) is called. I would expect all geometry data to be disposed of (and therefore listeners removed) when the renderer is dispose, as well. If those listeners are kept around after the renderer is done with then whatever is used in the scope of that function presumably can't be garbage collected, which includes the gl context.

**Comment by Mugen87** [comment]
The entire `dispose` system does not account for the use case when multiple renderers are used.

**Comment by gkjohnson** [comment]
> The entire dispose system does not account for the use case when multiple renderers are used.

That seems like a bug then, no? At least it would seem unintended from a users perspective. The docs state that `WebGLRenderer.dispose` will "Dispose of the current rendering context" but in some circumstances it seems it will not. You don't even necessarily need multiple renderers -- just a dangling reference to a geometry that was rendered at some point with the renderer you disposed:

```js
const renderer = new THREE.WebGLRenderer();
const geom = new THREE.BoxBufferGeometry();
const mat = new THREE.MeshBasicMaterial();
const mesh = new THREE.Mesh( geom, mat );
const camera = new THREE.PerspectiveCamera();

console.log( geom._listeners[ 'dispose' ].length, mat._listeners[ 'dispose' ].length ); // 0 0

// dispose listeners are added that include the gl context in the scope of the function
renderer.render( mesh, camera );

console.log( geom._listeners[ 'dispose' ].length, mat._listeners[ 'dispose' ].length ); // 1 1

// most objects are disposed of but listeners remain
renderer.dispose();

console.log( geom._listeners[ 'dispose' ].length, mat._listeners[ 'dispose' ].length ); // 1 1

// listeners are removed finally releasing the context reference but will also require disposing of the
// geometry in any other renderers
geom.dispose();
mat.dispose();

console.log( geom._listeners[ 'dispose' ].length, mat._listeners[ 'dispose' ].length ); // 0 0
```

I see that it's a complicated issue because WeakMaps (which have their own benefits) are used for geometry references which don't allow for iterating over the elements but on the surface it does seem to prevent `WebGLRenderer.dispose` from doing what it promises. Perhaps those docs should be clarified if this is intended behavior? I certainly didn't know I had to call dispose on the renderer _and_ all geometry and materials in order to fully clean things up.

**Comment by Mugen87** [comment]
> dispose on the renderer and all geometry and materials in order to fully clean things up.

Yes, this is intended. You've already pointed out the reason (non-iterable WeakMaps) for why this is the case. 

> Perhaps those docs should be clarified if this is intended behavior? 

Improving the docs is always good!

**Comment by gkjohnson** [comment]
I guess one question is whether or not there's another way to approach this that affords using two renderers at once or will it always be impossible to fully dispose of one renderer without incurring a reinitialization cost of shared geometry in the second.

One approach would be to enable disposing of objects directly on the renderer itself rather than with a dispose function:

```js
renderer.disposeObject( geometry );
```

Or of course using a `Map` rather than `WeakMap` in WebGLGeometries. What is the benefit of using a WeakMap right now? It will let the geometry objects be garbage collected in JS but it will never allow for the [gl bindings](https://github.com/mrdoob/three.js/blob/4944e56c06ed284abe11b7654348fb1ea3ac04d4/src/renderers/webgl/WebGLGeometries.js#L40) to be disposed of which will leave [unreferenced VAOs](https://github.com/mrdoob/three.js/blob/4944e56c06ed284abe11b7654348fb1ea3ac04d4/src/renderers/webgl/WebGLBindingStates.js#L472-L496) in memory until the renderer is disposed of.

I guess it seems like there's never a case where a user should feel comfortable with letting a BufferGeometry get garbage collected without calling dispose except for maybe just using a single renderer and having just disposed of the renderer. But even then they have to be sure that _all_ of the geometry is left unreferenced or the renderer context can't actually get cleaned up. If users are being told to call "dispose" on something it should actually properly dispose of the object. Are there strong feelings against using a `Map` rather than `WeakMap` in WebGLGeometries (and similar classes)?

Maybe WeakRef will make this all a big simpler but we'll see. In the mean time I can look into clarifying the wording on WebGLRenderer.dispose.
